File: packages/sdk/src/rtl/apiSettings.ts
/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { agentPrefix, defaultTimeout, ITransportSettings } from './transport'
import {
  lookerVersion,
  boolDefault,
  environmentPrefix,
  isTrue,
  unquote,
} from './constants'
import { IApiSection } from './nodeSettings'

export interface IValueSettings {
  [name: string]: string
}

export const strLookerBaseUrl = `${environmentPrefix}_BASE_URL`
export const strLookerVerifySsl = `${environmentPrefix}_VERIFY_SSL`
export const strLookerTimeout = `${environmentPrefix}_TIMEOUT`
export const strLookerClientId = `${environmentPrefix}_CLIENT_ID`
export const strLookerClientSecret = `${environmentPrefix}_CLIENT_SECRET`

export const ApiConfigMap: IValueSettings = {
  base_url: strLookerBaseUrl,
  client_id: strLookerClientId,
  client_secret: strLookerClientSecret,
  timeout: strLookerTimeout,
  verify_ssl: strLookerVerifySsl,
}

export const strBadConfiguration = `${agentPrefix} configuration error:
Missing required configuration values like base_url
`

export interface IApiSettings extends ITransportSettings {
  /**
   * Reading API settings on demand from some configuration source
   */
  readConfig(section?: string): IApiSection

  /**
   * Checks to see if minimal configuration values are assigned.
   *
   * If this function returns `false`, the Api configuration class will typically throw a run-time
   * error so the implementer knows required configuration values are missing
   *
   * @returns true or false
   */
  isConfigured(): boolean
}

/**
 * default the runtime configuration settings
 * @constructor
 *
 */
export const DefaultSettings = () =>
  ({
    agentTag: `${agentPrefix} ${lookerVersion}`,
    base_url: '',
    timeout: defaultTimeout,
    verify_ssl: true,
  } as IApiSettings)

/**
 * Return environment variable name value first, otherwise config name value
 * @param {IValueSettings} values
 * @param {string} name
 * @returns {string}
 */
export const configValue = (values: IValueSettings, name: string) => {
  const val = values[ApiConfigMap[name]] || values[name]
  return typeof val === 'string' ? unquote(val) : val
}

/**
 * Read any key/value collection for environment variable names and return as IApiSettings
 * @constructor
 *
 * The keys for the values are:
 *  - <environmentPrefix>_BASE_URL or `base_url`
 *  - <environmentPrefix>_CLIENT_ID or `client_id`
 *  - <environmentPrefix>_CLIENT_SECRET or `client_secret`
 *  - <environmentPrefix>_VERIFY_SSL or `verify_ssl`
 *  - <environmentPrefix>_TIMEOUT or `timeout`
 */
export const ValueSettings = (values: IValueSettings): IApiSettings => {
  const settings = DefaultSettings()
  settings.base_url = configValue(values, 'base_url') || settings.base_url
  settings.verify_ssl = boolDefault(configValue(values, 'verify_ssl'), true)
  settings.agentTag = `${agentPrefix} ${lookerVersion}`
  const timeout = configValue(values, 'timeout')
  settings.timeout = timeout ? parseInt(timeout, 10) : defaultTimeout
  return settings
}

/**
 * @class ApiSettings
 *
 * .ini Configuration initializer
 */
export class ApiSettings implements IApiSettings {
  base_url = ''
  verify_ssl = true
  timeout: number = defaultTimeout
  agentTag = `${agentPrefix} ${lookerVersion}`

  constructor(settings: Partial<IApiSettings>) {
    // coerce types to declared types since some paths could have non-conforming settings values
    this.base_url =
      'base_url' in settings ? unquote(settings.base_url) : this.base_url
    this.verify_ssl =
      'verify_ssl' in settings
        ? isTrue(unquote(settings.verify_ssl?.toString()))
        : this.verify_ssl
    this.timeout =
      'timeout' in settings
        ? parseInt(unquote(settings.timeout?.toString()), 10)
        : this.timeout
    if (!this.isConfigured()) {
      throw new Error(strBadConfiguration)
    }
  }

  isConfigured() {
    return !!this.base_url
  }

  /**
   * Default dynamic configuration reader
   * @param _section key/name of configuration section to read
   * @returns an empty `IAPISection`
   */
  readConfig(_section?: string): IApiSection {
    return {}
  }
}

File: examples/swift/sample-swift-sdk/sample-swift-sdk/LookerSwiftSDK/apiSettings.swift
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2019 Looker Data Sciences, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

import Foundation

let strLookerBaseUrl = "\(Constants.environmentPrefix)_BASE_URL"
let strLookerApiVersion = "\(Constants.environmentPrefix)_API_VERSION"
let strLookerVerifySsl = "\(Constants.environmentPrefix)_VERIFY_SSL"
let strLookerTimeout = "\(Constants.environmentPrefix)_TIMEOUT"
let strLookerClientId = "\(Constants.environmentPrefix)_CLIENT_ID"
let strLookerClientSecret = "\(Constants.environmentPrefix)_CLIENT_SECRET"
let strBadConfiguration = """
\(agentTag) configuration error:
Missing required configuration values like base_url and api_version
"""

protocol IApiSettings: ITransportSettings {
    func isConfigured() -> Bool
    func readConfig(_ section: String?) -> IApiSection
}

typealias IApiSection = SectionConfig

/**
 * default the runtime configuration settings
 * @constructor
 *
 */
struct DefaultSettings : IApiSettings {
    func isConfigured() -> Bool {
        return false
    }
    func readConfig(_ section: String? = nil) -> IApiSection {
        return [:]
    }
    var base_url: String? = ProcessInfo.processInfo.environment[strLookerBaseUrl] ?? ""
    var api_version: String? = ProcessInfo.processInfo.environment[strLookerApiVersion] ?? "4.0"
    var verify_ssl: Bool? = true
    var timeout: Int? = defaultTimeout
    var headers: Headers?
    var encoding: String?
}

/**
 * Read any key/value collection for environment variable names and return as IApiSettings
 * @constructor
 *
 * The values keys are:
 *  - <environmentPrefix>_BASE_URL
 *  - <environmentPrefix>_API_VERSION
 *  - <environmentPrefix>_CLIENT_ID
 *  - <environmentPrefix>_CLIENT_SECRET
 *  - <environmentPrefix>_VERIFY_SSL
 *  - <environmentPrefix>_TIMEOUT
 */
func ValueSettings(_ values: StringDictionary<String>) -> IApiSettings {
    var settings = DefaultSettings()
    settings.api_version = values[strLookerApiVersion] ?? settings.api_version
    settings.base_url = values[strLookerBaseUrl] ?? settings.base_url
    if (values[strLookerVerifySsl] != nil) {
        let v = values[strLookerVerifySsl]!.lowercased()
        settings.verify_ssl = v == "true" || v == "1"
    }
    if (values[strLookerTimeout] != nil) {
        settings.timeout = Int(values[strLookerTimeout]!)!
    }
    return settings
}

/**
 * @struct  ApiSettings
 *
 * .ini Configuration initializer
 */
struct ApiSettings: IApiSettings {
    func readConfig(_ section: String? = nil) -> IApiSection {
        return [:]
    }
    
    var base_url: String?
    var api_version: String?
    var verify_ssl: Bool?
    var timeout: Int?
    var headers: Headers?
    var encoding: String?
    
    init() {
        
    }
    
    init(_ settings: IApiSettings) throws {
        let defaults = DefaultSettings()
        // coerce types to declared types since some paths could have non-conforming settings values
        self.base_url = settings.base_url ?? defaults.base_url
        self.api_version = settings.api_version ?? defaults.api_version
        self.verify_ssl = settings.verify_ssl ?? defaults.verify_ssl
        self.timeout = settings.timeout ?? defaults.timeout
        self.headers = settings.headers ?? defaults.headers
        self.encoding = settings.encoding ?? defaults.encoding
        if (!self.isConfigured()) {
            throw SDKError(strBadConfiguration)
        }
    }
    
    func isConfigured() -> Bool {
        return (self.base_url != "" && self.api_version != "")
    }
}

